modules/widgets/instagram/instagram.php

```diff
@@ -0,0 +1,423 @@
+		// If removing the widget's stored token ID
+		elseif ( $instance['token_id'] && isset( $_GET['instagram_widget_id'] ) && $_GET['instagram_widget_id'] == $this->number && ! empty( $_GET['instagram_widget'] ) && 'remove_token' == $_GET['instagram_widget'] ) {
+			if ( empty( $_GET['nonce'] ) || ! wp_verify_nonce( $_GET['nonce'], 'instagram-widget-remove-token-' . $this->number . '-' . $instance['token_id'] ) ) {
+				wp_die( __( 'Missing or invalid security nonce.', 'wpcom-instagram-widget' ) );
+			}
+			$site = Jetpack_Options::get_option( 'id' );
+			$path = sprintf( '/sites/%s/instagram/%s/remove', $site, $instance['token_id'] );
+			$result = $this->wpcom_json_api_request_as_blog( $path, 2, array( 'headers' => array( 'content-type' => 'application/json' ) ), null, 'wpcom' );
+
+			$response_code = wp_remote_retrieve_response_code( $result );
+
+			if ( 200 !== $response_code ) {
+				do_action( 'wpcomsh_log', 'Instagram widget: failed to remove keyring token: API returned code ' . $response_code );
+				return 'ERROR';
+			}
+
+			$instance['token_id'] = $this->defaults['token_id'];
+
+			$this->update_widget_token_id( $instance['token_id'] );
+		}
+		// If a token ID is stored, make sure it's still valid
+		elseif ( $instance['token_id'] ) {
+			$this->update_widget_token_id( $instance['token_id'] );
+		}
+
+		// No connection, or a legacy API token? Display a connection link.
+		if ( ! $instance['token_id'] ) {
+			echo '<p>' . __( '<strong>Important: You must first click Save to activate this widget <em>before</em> connecting your account.</strong> After saving the widget, click the button below to authorize your Instagram account.', 'wpcom-instagram-widget' ) . '</p>';
+			?>
+			<script type="text/javascript">
+				function getScreenCenterSpecs( width, height ) {
+					const screenTop = typeof window.screenTop !== 'undefined' ? window.screenTop : window.screenY,
+						screenLeft = typeof window.screenLeft !== 'undefined' ? window.screenLeft : window.screenX;
+
+					return [
+						'width=' + width,
+						'height=' + height,
+						'top=' + ( screenTop + window.innerHeight / 2 - height / 2 ),
+						'left=' + ( screenLeft + window.innerWidth / 2 - width / 2 ),
+					].join();
+				};
+				function openWindow(button) {
+					window.open(
+						button.dataset.connecturl, //TODO: Check if this needs validation it could be a XSS problem. Check the domain maybe?
+						'_blank',
+						'toolbar=0,location=0,menubar=0,' + getScreenCenterSpecs( 700, 700 )
+					);
+					window.onmessage = function( { data } ) {
+						if ( !! data.keyring_id ) {
+							const payload = {
+								action: 'wpcom_instagram_widget_update_widget_token_id',
+								savetoken: '<?php echo esc_js( wp_create_nonce( 'instagram-widget-save-token' ) ); ?>',
+								keyring_id: data.keyring_id,
+							};
+							jQuery.post( ajaxurl, payload, function( response ) {
+								const widget = jQuery(button).closest('div.widget');
+								if ( ! window.wpWidgets ) {
```

Comment:
If we're in the customizer, then `wpWidgets` doesn't exist, but I also can't find a way of refreshing the widgets panel. It must be possible, but instead I've resorted to refreshing the whole page. This has implications, as the widget will need to have been published before the auth flow is initiated.